lib/src/sparql/mod.rs

```diff
@@ -121,70 +139,125 @@ impl<'a, S: StoreConnection + 'a> SimplePreparedQuery<S> {
+                options.service_handler,
+                options.custom_functions_handler,
+            ),
+        }))
     }
 }
 
 impl<S: StoreConnection> PreparedQuery for SimplePreparedQuery<S> {
-    fn exec<'a>(&'a self, options: &'a QueryOptions<'a>) -> Result<QueryResult<'a>> {
+    fn exec(&self) -> Result<QueryResult> {
         match &self.0 {
             SimplePreparedQueryAction::Select {
                 plan,
                 variables,
                 evaluator,
-            } => evaluator.evaluate_select_plan(&plan, &variables, options),
+            } => evaluator.evaluate_select_plan(&plan, &variables),
             SimplePreparedQueryAction::Ask { plan, evaluator } => {
-                evaluator.evaluate_ask_plan(&plan, options)
+                evaluator.evaluate_ask_plan(&plan)
             }
             SimplePreparedQueryAction::Construct {
                 plan,
                 construct,
                 evaluator,
-            } => evaluator.evaluate_construct_plan(&plan, &construct, &options),
+            } => evaluator.evaluate_construct_plan(&plan, &construct),
             SimplePreparedQueryAction::Describe { plan, evaluator } => {
-                evaluator.evaluate_describe_plan(&plan, &options)
+                evaluator.evaluate_describe_plan(&plan)
             }
         }
     }
 }
 
+/// Handler for SPARQL SERVICEs.
+///
+/// Might be used to implement [SPARQL 1.1 Federated Query](https://www.w3.org/TR/sparql11-federated-query/)
 pub trait ServiceHandler {
+    /// Get the handler for a given service identified by a RDF IRI.
+    ///
+    /// A service is a function that returns an iterator of bindings from a `GraphPattern`.
+    /// Returns `None` if there is no handler for the service.
     fn handle<'a>(
         &'a self,
-        node: NamedNode,
+        node: &NamedNode,
     ) -> Option<(fn(GraphPattern) -> Result<BindingsIterator<'a>>)>;
 }
 
+#[derive(Default)]
+struct EmptyServiceHandler {}
+
+impl ServiceHandler for EmptyServiceHandler {
+    fn handle(
+        &self,
+        _node: &NamedNode,
+    ) -> Option<(fn(GraphPattern) -> Result<BindingsIterator<'static>>)> {
+        None
+    }
+}
+
+/// Handler for custom functions.
+///
+/// For example, the following query uses a custom function to reverse a string:
+///
+///```sparql
+/// PREFIX ex: <http://example.com#>
+/// PREFIX foaf: <http://xmlns.com/foaf/0.1/>
+/// SELECT ?name ?reverse
+/// WHERE
+///   {
+///     ?s foaf:name ?name .
+///     BIND(ex:REVERSE(?name) as ?reverse)
+///   }
+/// ```
+pub trait CustomFunctionsHandler {
+    /// Handles all custom functions for a given query
+    ///
+    /// For the example above node would have the value http://example.com#REVERSE
+    /// and would recieve one parameter with the value mapped to ?name
+    ///
+    /// Returns `None` if there is an error or the given NamedNode isn't recognized
+    fn handle(&self, node: &NamedNode, parameters: &[Option<Term>]) -> Option<Term>;
```

Comment:
I would make the `parameters` type be `&[Term]`.

1. It would make implementations easier (just need to check the parameter count and not that all parameters are set).
2. Other implementations like [RDF4J](https://rdf4j.eclipse.org/documentation/custom-sparql-functions/) and [Jena](https://jena.apache.org/documentation/query/writing_functions.html) does not seem to allow failed evaluations as parameters.
3. We probably want expressions like `ex:fn(xsd:dateTime("foo"))` to fail instead of calling `ex:fn` on `[None]`.
4. Cases where an argument is in a maybe bound variable could be handled like `IF(BOUND(?optArg), ex:fn(?optArg), ex:fn())`.

To provide optional parameters, one could just set them at the end of the parameter list like what is done with the build in functions (e.g. [REGEX](https://www.w3.org/TR/sparql11-query/#func-regex)).

Having a signature: `fn handle(&self, node: &NamedNode) -> Option<(fn (parameters: &[Term]) -> Option<Term>)>` could be interesting to be consistent with the one for services, allow to distinguish the case where the function does not exists from the one where its evaluation failed and allow to call `CustomFunctionsHandler.handle` once per query evaluation if the function name is already known and just call the returned closure for each tuple. 

It could also make sense to change the return type from `Option<Term>` to `Result<Term>`. This way, if we want to implement a "debug" mode in the evaluator, the evaluator could raise meaningful errors from the custom functions without breaking change to the trait.

Reply:
> I would make the `parameters` type be `&[Term]`.
> 
>     1. It would make implementations easier (just need to check the parameter count and not that all parameters are set).
> 
>     2. Other implementations like [RDF4J](https://rdf4j.eclipse.org/documentation/custom-sparql-functions/) and [Jena](https://jena.apache.org/documentation/query/writing_functions.html) does not seem to allow failed evaluations as parameters.
> 
>     3. We probably want expressions like `ex:fn(xsd:dateTime("foo"))` to fail instead of calling `ex:fn` on `[None]`.
> 
>     4. Cases where an argument is in a maybe bound variable could be handled like `IF(BOUND(?optArg), ex:fn(?optArg), ex:fn())`.
> 

How are unbound values handled in rudf? I haven't run into them yet, but I assume this was represented by `None`. You can get these values with `OPTIONAL` clauses. Also it says in the SPARQL spec: [If the evaluation of the expression produces an error, the variable remains unbound for that solution but the query evaluation continues.](https://www.w3.org/TR/sparql11-query/#assignment), so it would seem to me at a minimum the `handle` function needs to return `Option<Term>`. If we follow your fourth suggestion, function composition couldn't really be done in this way: `ex:fn1(ex:fn2(1234))`. If unbound values are handled some other way, I guess it would work.

That aside, I don't think it's that bad ergonomically. Since the `handle` function returns `Option<Term>`, you can easily unbox parameters you expect to be there:

```rust
let param0 = paramaters[0].flatten()?; //assuming flatten is function on Option, 
let param1 = parameters[1].and_then(|p| p)?; // if flatten is something I imagined
```
That's some additional boiler plate, but not much worse than what it is without `Option`:

```rust
let param0 = paramaters[0]?;
let param1 = parameters[1]?;
```

> 
> To provide optional parameters, one could just set them at the end of the parameter list like what is done with the build in functions (e.g. [REGEX](https://www.w3.org/TR/sparql11-query/#func-regex)).
> 
> Having a signature: `fn handle(&self, node: &NamedNode) -> Option<(fn (parameters: &[Term]) -> Option<Term>)>` could be interesting to be consistent with the one for services, allow to distinguish the case where the function does not exists from the one where its evaluation failed and allow to call `CustomFunctionsHandler.handle` once per query evaluation if the function name is already known and just call the returned closure for each tuple.
> 

Ok, I can change this.

> It could also make sense to change the return type from `Option<Term>` to `Result<Term>`. This way, if we want to implement a "debug" mode in the evaluator, the evaluator could raise meaningful errors from the custom functions without breaking change to the trait.

This would require changing `eval_plan_expression`. Not really a problem but just an aside - that said see my comment about failed valued being unbound but the query continuing.

Reply:
> How are unbound values handled in rudf?

With `Option<EncodedTerm>`. It's why `EncodedTuple` is defined as `type EncodedTuple = Vec<Option<EncodedTerm>>`.

> so it would seem to me at a minimum the handle function needs to return Option<Term>.

Yes, I am not arguing about it, the custom functions needs to be able to fail.

> If we follow your fourth suggestion, function composition couldn't really be done in this way: ex:fn1(ex:fn2(1234)).

It could easily. The Rust evaluation code would look kinda like `self.custom_functions_handler.handle("ex:fn1", &[self.custom_functions_handler.handle("ex:fn2", &[1234])?])` (note the `?` after ex:fn2 evaluation).

My goal is mostly to follow what other SPARQL implementations are doing and the behavior of built in functions: if a function parameter evaluation returns an error (possibly because a variable is unbound), the function is not called and an error is returned instead. This way, we forbid the creation of "error caching" functions but it's probably better for interoperability with other implementations that do not allow it either. 

>> It could also make sense to change the return type from Option<Term> to Result<Term>. This way, if we want to implement a "debug" mode in the evaluator, the evaluator could raise meaningful errors from the custom functions without breaking change to the trait.
>
> This would require changing eval_plan_expression. Not really a problem but just an aside - that said see my comment about failed valued being unbound but the query continuing.

If we just make custom function return results, we don't have to change much: One could just write `self.custom_functions_handler.handle(name, &parameters).ok()` to get back an `Option<T>` from a `Result<T>` just like it's done with the numerous calls to Rust std library functions that returns `Result`.